improving docs for v0.3 + fix markdown backend

pietroppeter opened the pull request:
fixes #85 (last step before tagging and releasing)

- improve changelog
- make sure all important changes (recent and older) are documented

details:

- [x] improving changelog and integrating thanks in changelog
  - [x] 0.1
  - [x] 0.1.x
  - [x] 0.2
  - [x] 0.2.x
  - [x] 0.3
  - [x] move changelog to separate file
- [x] improve documentation for some of the recent changes (from 0.2.0 until now)
  - [x] stuff in 0.2.0 that went un(der)documented
    - [x] add list of command line options generated with `nbInit`
  - [x] changes in 0.2.x
    - [x] 0.2.2: nbFile
  - [x] other 0.3 improvements by Hugo #80
    - [x] nbRawOutput
    - [x] nbClearOutput
  - [x] release 0.3 stuff #81
    - [x] newNbCodeBlock
    - [x] newNbSlimBlock
    - [x] new nbTextWithCode that does read code
    - [x] example files.nim for File
  - [x] nbPython # 83
    - [x] nbInitPython and nbPython
  - [x] nim to javascript #88 - added a section
    - [x] nbCodeToJs (or by pieces nbInitToJs, addCodeToJs, addToDocAsJs)
    - [x] nbKaraxCode
  - [x] CodeAsInSource now the default:
    - [x] remove bullet point in API section
    - [x] add section "Code Capture"
  - [x] make sure to rerun nimble readme to have readme updated (currently missing the interactivity docs)
- [x] other changes
  - [x] turned off warning for unused imports in `nimib.nim` (#103)
  - [x] fix md output of index.nim (bug in markdown backend?)


- some other improvements that could be done later: https://github.com/pietroppeter/nimib/issues/110

pietroppeter commented:
ok, this is done. After this is merged I would tag and release.
@HugoGranstrom would you give it a look?

some small details:
- I did not document `nbClearOutput` because I cannot remember what was its use case.
- I removed random from import (I think it was used in js code generation but now it should not be anymore needed)
- I am not sure if it is worth adding at the end of nimib.nim a pragma to turn off unusedImport warning (a normal run of index.nim would still mark sequtils and strutils as not used). I also found very little documentation of this, basically only https://forum.nim-lang.org/t/5461#34187

HugoGranstrom commented:
> ok, this is done. After this is merged I would tag and release.
> @HugoGranstrom would you give it a look?

:confetti_ball: 
I'll take a closer look later tonight but I'll answer your questions:

> I did not document nbClearOutput because I cannot remember what was its use case.

It was mainly for nimiSlides to save some screen estate if the output of the code block wasn't important. (ggplotnim for example echo:ed some really noisy stuff for example)

> I removed random from import (I think it was used in js code generation but now it should not be anymore needed)

Ah yes forgot about that one. Also I saw that they have released a new Karax release so we should be able to drop the `@#head` in the `nimib.nimble` file.

> I am not sure if it is worth adding at the end of nimib.nim a pragma to turn off unusedImport warning (a normal run of index.nim would still mark sequtils and strutils as not used). I also found very little documentation of this, basically only https://forum.nim-lang.org/t/5461#34187

As long as it doesn't affect the user's imports, I think that would be a welcome addition :+1:

pietroppeter commented:
> > I did not document nbClearOutput because I cannot remember what was its use case.
> 
> It was mainly for nimiSlides to save some screen estate if the output of the code block wasn't important. (ggplotnim for example echo:ed some really noisy stuff for example)

ok, makes sense. added explanation.

> > I removed random from import (I think it was used in js code generation but now it should not be anymore needed)
> 
> Ah yes forgot about that one. Also I saw that they have released a new Karax release so we should be able to drop the `@#head` in the `nimib.nimble` file.

also update Karax version

> > I am not sure if it is worth adding at the end of nimib.nim a pragma to turn off unusedImport warning (a normal run of index.nim would still mark sequtils and strutils as not used). I also found very little documentation of this, basically only https://forum.nim-lang.org/t/5461#34187
> 
> As long as it doesn't affect the user's imports, I think that would be a welcome addition 👍

good idea. I checked and indeed it does not affect user's imports. Apparently it only affects import of the module where it is added. So added.

Instead I am having issues with the markdown backend. I did not realize it is actually using parts of html backend and not sure why... will look into that and fix it.

HugoGranstrom reviewed (changes requested):
Nothing major really, just a few nitpicks. It looks very good :rocket: 

But wow! There has been so much happening to nimib over the last year, I can't believe it was just over a year since v0.1 was released :exploding_head:

pietroppeter commented on docsrc/index.nim:
```diff
@@ -123,6 +127,41 @@ Customization over the default is mostly achieved through nim-mustache or changi
+  Example blocks created with `newNbCodeBlock` are `nbCode` and `nbTextWithCode`.
+* `newNbSlimBlock(cmd: string, blockImpl: untyped)`: template that can be used to create
+  a custom block that does not need a separate `body`.
+  Example blocks created with `newNbSlimBlock` are `nbText`, `nbImage` and `nbFile`.
```
you are right, will remove the ref to nbFile

pietroppeter commented on changelog.md:
```diff
@@ -0,0 +1,221 @@
+  - all nimib "native" blocks now use one of the two mechanism above
+  - now `nbText` does NOT contain its source
+  - new `nbTextWithCode` that does contain code source
+* untyped version of `nbFile` removed, new example document `files.nim`, changed the rendering of `nbFile`
```
indeed I do not know it it was removed at some point or not (I already forgot if I added it back in some previous PR)

pietroppeter commented:
> But wow! There has been so much happening to nimib over the last year, I can't believe it was just over a year since v0.1 was released 🤯

indeed when writing the history/changelog I was also surprised by how many changes there were!

fixed your remarks and also fixed the markdown backend which required some non trivial changes. Also updated the rendering documentation which was indeed obsolete.

HugoGranstrom commented:
> fixed your remarks and also fixed the markdown backend which required some non trivial changes. Also updated the rendering documentation which was indeed obsolete.

:rocket: :partying_face: Awesome! It felt very typical that it was the thing we will refactor next that was misbehaving, I'll take that as a sign :rofl: 

So all that's left now is to go through the preview one more time and then release 0.3? :exploding_head:

pietroppeter commented:
> So all that's left now is to go through the preview one more time and then release 0.3? 🤯

Yep! I will also do another review later on

HugoGranstrom commented on docsrc/index.nim:
```diff
@@ -1,17 +1,20 @@
 
 let
   repo = "https://github.com/pietroppeter/nimib"
   docs = if defined(useMdBackend): "https://pietroppeter.github.io/nimib" else: "."
```
```suggestion
  docs = if defined(mdOutput): "https://pietroppeter.github.io/nimib" else: "."
```
The readme links were incorrect

pietroppeter commented:
made another pass but seems good to go! 🥳

pietroppeter commented:
you are right, will do
